File: 020_factorial_digit_sum.php
<?php

/*\
| | Project #20
| |
| | n! means n * (n - 1) * ... * 3 * 2 * 1
| |
| | For example, 10! = 10 * 9 * ... * 3 * 2 * 1 = 3628800, and the sum
| | of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.
| |
| | Find the sum of the digits in the number 100!
\*/

// NOTE: PHP represents large number in scientific notation
// so this will not work for 100!

$n = 10;
$result = factorial($n);
$sum = sum_digits($result);

echo "$n! = " . factorial($n) . "\n";
echo "sum of digits = $sum\n";

function factorial($n) {
    if($n == 1) { return 1; }

    return factorial($n -1) * $n;
}

function sum_digits($n) {
    $s = (string)$n;
    $len = strlen($s);
    $sum = 0;
    for($i = 0; $i < $len; $i++) {
        $sum += (int)substr($s, $i, 1);
    }
    return $sum;
}

?>

File: 020_factorial_digit_sum.rb

# /*\
# | | Project #16
# | |
# | | 2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
# | |
# | | What is the sum of the digits of the number 2^1000?
# \*/

def factorial(n)
  (1..n).reduce(:*) || 1
end

n = 100

sum_of_digits = factorial(n).to_s.split('').inject(0) { |sum, i| sum + i.to_i }

puts "#{n}! = #{factorial(n)}"
puts "Sum of digits: #{sum_of_digits}"
